### PR TITLE
H3 4.0.0 Part 1: Renaming otherwise unchanged functions

### PR DESCRIPTION
**Pull request by dfellis**
Follows the renaming described here: https://github.com/uber/h3/blob/master/dev-docs/RFCs/v4.0.0/names_for_concepts_types_functions.md

### CODE REVIEW

**Comment by dfellis** [comment]
I believe this is ready for review!

**Comment by dfellis** [comment]
@sahrk I can't formally assign you as a reviewer, but in case you are interested, we're finally getting started on cleaning up the naming convention for a v4 release.

File: src/h3lib/lib/algos.c
Comment (nrabinowitz) [inline src/h3lib/lib/algos.c:441]: Nit: Should be `gridDiskDistancesUnsafe`, right?

File: src/h3lib/lib/directedEdge.c
Comment (nrabinowitz) [inline src/h3lib/lib/directedEdge.c:250]: This is always an issue with `isX` functions, that the output also seems like it should be called `isX`. I'd prefer a more specific name here, e.g. `originIsPentagon`, since we frequently use `pentagon` for a pentagon H3 index - though the easiest option is probably just to inline `H3_EXPORT(isPentagon)(origin)` in the `if` condition.

Is there a chance for subtle bugs here with `H3_EXPORT`, given our more generic names, or will this fail compilation?

File: src/h3lib/lib/faceijk.c
Comment (nrabinowitz) [inline src/h3lib/lib/faceijk.c:418]: I'm finding this one a little confusing, naming-wise. What about going the other way, and using `isClassIII` for testing the resolution? or, alternatively, using a `IS_CLASS_III` macro?

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:360]: This one looks like a find&replace error

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:368]: Ditto

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:395]: Ditto, here and below

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:499]: Some more F&R issues

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:851]: Not required, but we might want to change `gb` here and possibly elsewhere (likewise in the doc comment)

File: src/h3lib/lib/h3Index.c
Comment (nrabinowitz) [inline src/h3lib/lib/h3Index.c:886]: Again, I don't really like `pentagon` here - maybe `isCellPentagon`? I wish there was a nicer rule here.

File: src/h3lib/lib/vertex.c
Comment (nrabinowitz) [inline src/h3lib/lib/vertex.c:115]: Ditto, maybe `isOriginPentagon`?

**Review by nrabinowitz — COMMENTED** [review]
Oof! A few comments, but overall this looks great - thanks for taking this on!

File: examples/compact.c
Comment (dfellis) [reply examples/compact.c:38]: They use H3 as a library, so `H3_EXPORT` is not available, and the test compiles it with an empty prefix so this works (or the test suite would be complaining ;) )

File: CMakeLists.txt
Comment (dfellis) [reply CMakeLists.txt:185]: I couldn't find any evidence anywhere that it was being renamed or having its experimental status removed, so that's why I didn't touch it in this PR. This *is* just part 1, so we can revisit this :)

File: src/apps/applib/lib/utility.c
Comment (dfellis) [reply src/apps/applib/lib/utility.c:102]: Sure.

File: src/apps/applib/lib/utility.c
Comment (dfellis) [reply src/apps/applib/lib/utility.c:358]: :+1:

File: src/apps/filters/pointToCell.c
Comment (dfellis) [reply src/apps/filters/pointToCell.c:20]: Should definitely get that decided before we land this.

I would really like to know if anyone actually uses these utility programs, though.

File: src/apps/filters/h3ToComponents.c
Comment (dfellis) [reply src/apps/filters/h3ToComponents.c:53]: Probably, but I wasn't sure if that behavior should be changed, or even if we're gonna keep these programs around.

File: src/apps/testapps/testGridDisksUnsafe.c
Comment (dfellis) [reply src/apps/testapps/testGridDisksUnsafe.c:22]: I don't think we do. We got the code coverage for both through this test and left it there, but you're right that that's an implementation detail that the test suite theoretically shouldn't worry about. Can we create an issue for that? I think it's out of scope of this massive PR.

File: src/apps/testapps/testGridPathCells.c
Comment (dfellis) [reply src/apps/testapps/testGridPathCells.c:19]: I'll let you two hash that out. I don't have a strong view either way on this.

File: src/h3lib/include/coordijk.h
Comment (dfellis) [reply src/h3lib/include/coordijk.h:2]: Well, there weren't any changes to copyright in 2019, so this is more accurate.

File: src/h3lib/lib/algos.c
Comment (dfellis) [reply src/h3lib/lib/algos.c:246]: That's reasonable.

File: src/h3lib/lib/algos.c
Comment (dfellis) [reply src/h3lib/lib/algos.c:441]: Should it? I though the document just said "gridDiskUnsafe"?

File: src/h3lib/lib/directedEdge.c
Comment (dfellis) [reply src/h3lib/lib/directedEdge.c:250]: It will fail compilation, but only if the prefix is not set (which is the default).

I went with this name for the variable because it was easier to write a vim macro to change all instances across the source files automatically.

File: src/h3lib/lib/faceijk.c
Comment (dfellis) [reply src/h3lib/lib/faceijk.c:418]: So `isResClassIII` was taken over by `h3IsResIII` (or something like that) and I had to rename this function. I didn't want to refactor it into a macro or anything because it would make proving that the renaming was correct even more difficult than it already is in a code review. Having this renaming PR as straightforward as possible makes reviewing the monstrosity more manageable.

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:360]: That it is. Will fix.

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:368]: Thanks.

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:395]: :+1:

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:499]: :+1:

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:851]: I can do that.

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:886]: I'd rather not change this one, though.

File: src/h3lib/lib/vertex.c
Comment (dfellis) [reply src/h3lib/lib/vertex.c:115]: Same comment.

File: src/apps/testapps/testGridDisksUnsafe.c
Comment (nrabinowitz) [reply src/apps/testapps/testGridDisksUnsafe.c:22]: Agreed, added #407

File: src/h3lib/lib/algos.c
Comment (nrabinowitz) [reply src/h3lib/lib/algos.c:441]: Unless I'm reading the diff wrong, `gridDiskUnsafe` is the function above this, and this comment refers to `gridDiskDistancesUnsafe` below

File: src/h3lib/include/coordijk.h
Comment (nrabinowitz) [reply src/h3lib/include/coordijk.h:2]: 🤷‍♂️

File: src/h3lib/lib/directedEdge.c
Comment (nrabinowitz) [reply src/h3lib/lib/directedEdge.c:250]: I guess my point is partly that there's no way to change it _back_ now, because `pentagon` is widely used as a variable name for something else in other contexts. Up to you, but I find `pentagon` confusing here

File: src/h3lib/lib/faceijk.c
Comment (nrabinowitz) [reply src/h3lib/lib/faceijk.c:418]: Ok, stet

File: src/apps/applib/lib/utility.c
Comment (ajfriend) [reply src/apps/applib/lib/utility.c:358]: Hmm. This function is tricky. I suppose the *intention* is that it would only be used on arrays of cells, but it would also work for arrays of other modes (and would even include "invalid" indexes other than `H3_NULL` in the count).

And the "`Hexagons`" in the function name should at least be "`Cells`", if not the more general `"Indexes"`.

What would you all think about a function signature like:

```c
/**
 * Returns the number of indexes in the array which are not H3_NULL
 */
int countNonNullIndexes(H3Index* indexes, int numIndexes) {
```

File: src/apps/filters/pointToCell.c
Comment (nrabinowitz) [reply src/apps/filters/pointToCell.c:20]: Discussed in the steering committee. @isaacbrodsky suggests consolidating these in a single filter command like `h3 pointToCell <...args>`. This would be a separate PR, obviously, so what you have here seems sufficient for now.

File: src/apps/filters/h3ToComponents.c
Comment (ajfriend) [reply src/apps/filters/h3ToComponents.c:53]: +1 to changing this to `"Cell"`

File: src/apps/testapps/testGridPathCells.c
Comment (ajfriend) [reply src/apps/testapps/testGridPathCells.c:19]: How about we just call this `gridPath` with (`Cell` being implicit), while still having `gridPathEdges` and `gridPathDirectedEdges` (if we ever do write those functions).

I like `gridPath` more than `gridLine`, just because it aligns with a "path" in graph theory. If I didn't know what this function did, I think I'd be more likely to guess correctly with `gridPath`, but that's just my take. I'm not sure what the "average" user would guess.

File: src/h3lib/lib/directedEdge.c
Comment (ajfriend) [reply src/h3lib/lib/directedEdge.c:250]: Maybe `int isPent = H3_EXPORT(isPentagon)(origin);`?

File: src/h3lib/include/h3api.h.in
Comment (ajfriend) [inline src/h3lib/include/h3api.h.in:309]: Actually, we should just call this `getEdgeLengthAvg`, since I'll be computing the new average edge length values over **all** edges, without excluding any that might be incident with a pentagon.

File: src/h3lib/include/h3api.h.in
Comment (ajfriend) [inline src/h3lib/include/h3api.h.in:324]: These should just be `edgeLengthRads`, `edgeLengthKm`, and `edgeLengthM`, since we can drop the `exact` because we no longer have a naming conflict with `getEdgeLengthAvg`.

File: CMakeLists.txt
Comment (ajfriend) [reply CMakeLists.txt:185]: Agree that we'd probably want `cellToLocalIj`, but I think I skipped over it in the naming doc because of the experimental status. @isaacbrodsky, were we ever planning to graduate this out of "experimental"?

File: src/h3lib/lib/h3Index.c
Comment (ajfriend) [reply src/h3lib/lib/h3Index.c:886]: As above, maybe `int isPent = H3_EXPORT(isPentagon)(h3);`?

File: src/apps/testapps/testGridPathCells.c
Comment (nrabinowitz) [reply src/apps/testapps/testGridPathCells.c:19]: Sure, I think that's an improvement.

File: src/h3lib/include/h3api.h.in
Comment (isaacbrodsky) [inline src/h3lib/include/h3api.h.in:488]: Was `areCellsNeighbors` considered?

File: src/h3lib/lib/faceijk.c
Comment (isaacbrodsky) [reply src/h3lib/lib/faceijk.c:418]: I would suggest renaming to `isResolutionClassIII` or similar, there is no use of indexing digits ehre. I understand the concern with renaming to `isResClassIII`, which would be my preferred name.

File: src/h3lib/lib/h3Index.c
Comment (isaacbrodsky) [inline src/h3lib/lib/h3Index.c:536]: Seems like this should accept a resolution rather than an index.

**Review by isaacbrodsky — COMMENTED** [review]
I apologize for the delay in reviewing this PR. Thanks for pushing this part of the 4.0 effort forward.

I was able to complete reading through it and I think it makes sense to me. I see Nick and AJ have already commented on a few things, and I agree especially with the gridDiskDistancesUnsafe comment (or the one where the arguments were more complicated than needed), etc.

File: src/h3lib/lib/h3Index.c
Comment (dfellis) [reply src/h3lib/lib/h3Index.c:536]: So this is letting people know if the hexagon is a class III hexagon. I would prefer it to just be `isClassIII` and return `isResDigitClassIII` to this name, but that isn't what we agreed upon in the RFC and I think the relitigation of these decisions all the time is why 4.0 is taking so long to get out the door. :/

Comment (isaacbrodsky) [reply src/h3lib/lib/h3Index.c:536]: This function name is probably fine to keep. I don't think it blocks this PR

**Comment by sahrk** [comment]
> @sahrk I can't formally assign you as a reviewer, but in case you are interested, we're finally getting started on cleaning up the naming convention for a v4 release.

Thanks for the heads up @dfellis, and sorry for being late to the party. I am looking at it now and will make any comments this evening. Overall it looks great; I'm especially happy to have a standard naming for "ring" and "disk" that can be adopted outside of H3 going forward.

File: src/apps/filters/pointToCell.c
Comment (sahrk) [reply src/apps/filters/pointToCell.c:20]: @dfellis I use these :). I can't be the only one that uses pipes in the unix shell to transform data.

I concur with @nrabinowitz that the "CLI utils is one place where the more generic names aren't helpful". Within H3 itself the decision was made to make lat/lon equivalent to `point`, etc. But the command line user is outside H3 trying to transform between specific location representations, and using the abstract terms obscures that.

**Comment by sahrk** [comment]
Within the function names `Point` is now synonymous with lat/lon. It feels a little redundant to me to then use `GeoPoint` for the lat/lon data type. Maybe it should just be `Point`?

File: CMakeLists.txt
Comment (dfellis) [reply CMakeLists.txt:185]: Based on the last meeing, going to leave this one alone for this PR

File: src/apps/testapps/testGridPathCells.c
Comment (dfellis) [reply src/apps/testapps/testGridPathCells.c:19]: We can have a followup PR on this including the RFC change?

**Review by isaacbrodsky — APPROVED** [review]

File: src/apps/testapps/testGridPathCells.c
Comment (ajfriend) [reply src/apps/testapps/testGridPathCells.c:19]: Will do!

Comment (ajfriend) [reply src/apps/testapps/testGridPathCells.c:19]: https://github.com/uber/h3/pull/441

I also added an open question around if we should do a similar thing for `compact` and `uncompact`.

File: src/h3lib/include/h3api.h.in
Comment (ajfriend) [reply src/h3lib/include/h3api.h.in:338]: @dfellis bump :)

Comment (ajfriend) [reply src/h3lib/include/h3api.h.in:338]: Actually, never mind. I see that these aren't in the RFC. I'll make an issue for a follow-up PR to change these.

Comment (ajfriend) [reply src/h3lib/include/h3api.h.in:338]: https://github.com/uber/h3/issues/442

File: src/apps/testapps/testGridPathCells.c
Comment (ajfriend) [reply src/apps/testapps/testGridPathCells.c:19]: I'd vote for you to land this PR as-is, and I can take a stab at implementing whatever we decide for #441 in a follow-up PR.

File: src/h3lib/lib/h3Index.c
Comment (ajfriend) [reply src/h3lib/lib/h3Index.c:541]: I'm happy to revisit in a follow-up. I could debate names all day 😛 .

I think these changes have taken so long just due to the scale and the coupling between name changes. After this PR lands, I think @dfellis has done most of the hard work. I'd expect follow-up adjustments like this to be smaller-scale and less-coupled, so we should be able to resolve them fairly quickly. 🙏

File: CMakeLists.txt
Comment (ajfriend) [reply CMakeLists.txt:162]: @isaacbrodsky Could you make a follow-up issue? Or is this covered by https://github.com/uber/h3/issues/415?

**Comment by ajfriend** [comment]
> Within the function names `Point` is now synonymous with lat/lon. It feels a little redundant to me to then use `GeoPoint` for the lat/lon data type. Maybe it should just be `Point`?

~~Good point.~~ Fair. Let's consider in follow-up? https://github.com/uber/h3/issues/415

**Review by ajfriend — APPROVED** [review]
Looks great! Thanks for doing all the hard work @dfellis! 

I've tried to move any outstanding discussions to issues/PRs that we can follow up on, so that we can finally land this thing! 🎉

**Pull request merged**
